libs/execution_context/ebpf_maps.c

```diff
@@ -2064,33 +2111,298 @@ _update_circular_map_entry(
-        ebpf_ring_buffer_map_async_query_result_t* async_query_result = context->async_query_result;
-        size_t consumer;
-        ebpf_ring_buffer_query((ebpf_ring_buffer_t*)map->data, &consumer, &async_query_result->producer);
-        if (consumer > async_query_result->consumer) {
-            // We will normally already have the latest consumer value - only update if we see an increase.
-            async_query_result->consumer = consumer;
-        }
+    while (!ebpf_list_is_empty(&async_contexts->contexts)) {
+        ebpf_core_map_async_query_context_t* context =
+            EBPF_FROM_FIELD(ebpf_core_map_async_query_context_t, entry, async_contexts->contexts.Flink);
+        ebpf_map_async_query_result_t* async_query_result = context->async_query_result;
+        table->query_ring_buffer(map, context->index, async_query_result);
         ebpf_list_remove_entry(&context->entry);
-        ebpf_operation_ring_buffer_map_async_query_reply_t* reply =
-            EBPF_FROM_FIELD(ebpf_operation_ring_buffer_map_async_query_reply_t, async_query_result, async_query_result);
+
+        ebpf_operation_map_async_query_reply_t* reply =
+            EBPF_FROM_FIELD(ebpf_operation_map_async_query_reply_t, async_query_result, async_query_result);
         ebpf_async_complete(context->async_context, sizeof(*reply), EBPF_SUCCESS);
         ebpf_free(context);
         context = NULL;
     }
+    EBPF_LOG_EXIT();
+}
+
+static void
+_map_async_query_cancel(
+    _Inout_ ebpf_core_map_async_contexts_t* async_contexts, _In_ ebpf_core_map_async_query_context_t* context)
+{
+    ebpf_lock_state_t state = ebpf_lock_lock(&async_contexts->lock);
+    ebpf_list_remove_entry(&context->entry);
+    ebpf_lock_unlock(&async_contexts->lock, state);
+    ebpf_async_complete(context->async_context, 0, EBPF_CANCELED);
+    ebpf_free(context);
+}
+
+static void
+_query_ring_buffer_map(
+    _In_ const ebpf_core_map_t* map, uint64_t index, _Inout_ ebpf_map_async_query_result_t* async_query_result)
+{
+    UNREFERENCED_PARAMETER(index);
+    ebpf_ring_buffer_query(
+        (ebpf_ring_buffer_t*)map->data, &async_query_result->consumer, &async_query_result->producer);
+}
+
+static void
+_query_perf_event_array_map(
+    _In_ const ebpf_core_map_t* map, uint64_t index, _Inout_ ebpf_map_async_query_result_t* async_query_result)
+{
+    ebpf_core_perf_event_array_map_t* perf_event_array_map =
+        EBPF_FROM_FIELD(ebpf_core_perf_event_array_map_t, core_map, map);
+    ebpf_core_perf_ring_t* ring = &perf_event_array_map->rings[(uint32_t)index];
+    ebpf_ring_buffer_query(&ring->ring, &async_query_result->consumer, &async_query_result->producer);
+    async_query_result->lost_count = ring->lost_records;
+}
+
+static void
+_ebpf_ring_buffer_map_cancel_async_query(_In_ _Frees_ptr_ void* cancel_context)
+{
+    EBPF_LOG_ENTRY();
+    ebpf_core_map_async_query_context_t* context = (ebpf_core_map_async_query_context_t*)cancel_context;
```

Comment:
since the cancel context is cast to the same type `ebpf_core_map_async_query_context_t*` both here and in line 2273, why not change the parameter type to `ebpf_core_map_async_query_context_t*` ?